implement managed fs support

Ocawesome101 opened the pull request:
This PR is for tracking progress on managed-filesystem support.

- [x] attribute loading/saving
- [x] file access
- [x] `exists()` and `stat()`
- [x] `chmod()` and `chown()`
- [x] `link()` and `unlink()`
- [x] `mkdir()`
- [x] `opendir()` and `readdir()`
- [x] `read()`, `write()`, `seek()`, `flush()`, `close()`

Ocawesome101 committed:
initial single line of code

Ocawesome101 committed:
stubs

Ocawesome101 committed:
add argument stuff

Ocawesome101 committed:
add file to gitignore

Ocawesome101 committed:
lol mistyped name

Ocawesome101 committed:
attribute load/save, i think

atirutw committed:
Try to make codes more readable

I also think we should void single line "if this return nil and error" statements but I'll leave that for a later discussion.

atirutw commented on src/fs/managed.lua:
```diff
@@ -32,18 +32,74 @@ do
+      end
+    end
+    return data
+  end
```
Is this really necessary? It's not like `load_attributes()` cares about the ordering. Perhaps you intend to only dump whitelisted keys?

Ocawesome101 replied:
It really probably isn't. I'll remove it.

atirutw commented on src/fs/managed.lua:
```diff
@@ -32,18 +32,74 @@ do
+    local attributes = {}
+    for line in data:gmatch("[^\n]+") do
+      local key, val = line:match("^(.-):(.+)$")
+      attributes[key] = tonumber(val)
```
Is it possible that a corrupted attribute file could cause major errors? Since if a key is missing it'd be nil.

Ocawesome101 replied:
Possibly - I can add some sanity checks / field defaults.

Ocawesome101 committed:
improve attribute handling

Ocawesome101 committed:
whoops, had an unnecessary bitop

atirutw committed:
Fix typo

Ocawesome101 committed:
implement :exists() and :stat()

Ocawesome101 committed:
Merge branch 'managed-fs' of github.com:ocawesome101/oc-cynosure-2 into managed-fs

atirutw committed:
Merge remote-tracking branch 'origin/dev' into managed-fs

Ocawesome101 committed:
fix wrong variable name in urls/main causing :open() to break

atirutw committed:
Merge remote-tracking branch 'origin/managed-fs' into managed-fs

Ocawesome101 committed:
register the fs, probably

Ocawesome101 committed:
oops, oversight

Ocawesome101 committed:
lol, typo

Ocawesome101 committed:
another typo

atirutw committed:
Implement `open()`

Ocawesome101 committed:
add arg checking for is_attribute

Ocawesome101 committed:
maybe fix?

Ocawesome101 committed:
well.  that was fun.

atirutw committed:
Fix not being able to create new file

Ocawesome101 committed:
add 'fds' field to processes

atirutw committed:
Implement `write` and `close`

atirutw committed:
Implement some more functions

- `chmod` and `chown`
- `mkdir`
- `read`
- `seek`

atirutw committed:
Remove redundant `checkArg`s

Ocawesome101 committed:
forgot to return stat info; ensure chmod() can't set the file type bit of a file mode; makeDirectory cannot create "attribute files"

Ocawesome101 committed:
Merge branch 'managed-fs' of github.com:ocawesome101/oc-cynosure-2 into managed-fs

Ocawesome101 committed:
re-add argument checking.

@atirut-w arg checking is *always* a good thing - having it here helps
spot errors further up the stack that we might otherwise make without
noticing

Ocawesome101 committed:
i think this might actually be done

atirutw reviewed (approved):
Looks good so far
